p2p/src/supervisor.rs

```diff
@@ -0,0 +1,90 @@
+use crate::peer::{self, Peer};
+use crate::transport::{self, Connection, Endpoint, Transport};
+
+#[derive(Debug, thiserror::Error)]
+enum Error {
+    #[error(transparent)]
+    Transport(#[from] transport::Error),
+}
+
+trait State {}
+struct Stopped;
+impl State for Stopped {}
+struct Running<E, I>
+where
+    E: Endpoint,
+    I: Iterator<Item = Result<E::Connection, transport::Error>>,
+{
+    endpoint: E,
+    incoming: I,
+}
+impl<E, I> State for Running<E, I>
+where
+    E: Endpoint,
+    I: Iterator<Item = Result<E::Connection, transport::Error>>,
+{
+}
+
+struct Supervisor<T, St>
```

Comment:
how does supervisor learns about disconnects?

Reply:
Valid question, needs to be determined.